File: runme/mcgen/local/datan/exasrc/e1ml.f
      PROGRAM E1ML
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION T(1000),XPL(500),YPL(500)
      CHARACTER*75 TX,TY,CAPT
      PARAMETER(FACT=.39894228D0)
      EXTERNAL FUE1ML
      DOUBLE PRECISION FUE1ML
      DATA TX/'t,&t'/,TY/'-(l-l_max#)'/
C identify program to user
      WRITE(*,*)' Program E1ML computes mean life and'
      WRITE(*,*)' asymmetric errors from few radioactive decays'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number N of decays (1 <= N <= 1000)'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)' Enter seed number ISEED1 (>0)'
      WRITE(*,*)' >'
      READ(*,*)ISEED1
      WRITE(*,*)' Enter seed number ISEED2 (>0)'
      WRITE(*,*)' >'
      READ(*,*)ISEED2
C initialization
      IF(ISEED1.LE.0) ISEED1=123456
      IF(ISEED2.LE.0) ISEED2=654321
      CALL RNE2IN(ISEED1,ISEED2)
C prepare graphics
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      CALL GROPEN
      CALL GROPWS(NWS)
      CALL GRWNCC(0.D0,5.D0,0.D0,0.6D0)
      CALL GRVWWC(0.1D0,1.1D0,0.1D0,0.75D0)
      CALL GRWNWC(-0.214D0,1.2D0,-0.1D0,0.9D0)
      CALL GRSTFR(NWS,5.D0,0.D0)
      CALL GRFRAM
      CALL GRSCLX(TX,4)
      CALL GRSCLY(TY,11)
      CALL GRBOUN
C generate sample and draw 1D scatter diagram
      CALL RNECUY(T,N)
      TBAR=0.D0
      YPL(1)=0.0D0
      YPL(2)=0.1D0
      CALL GRSTCL(2)
      DO 2 I=1,N
        T(I)=-LOG(T(I))
        TBAR=TBAR+T(I)
        XPL(1)=T(I)
        XPL(2)=T(I)
        CALL GRPLIN(2,XPL,YPL)
2     CONTINUE
C draw line -(l-lmax)=.5
      TBAR=TBAR/DBLE(N)
      XPL(1)=0.D0
      XPL(2)=5.D0
      YPL(1)=.5D0
      YPL(2)=.5D0
      CALL GRPLIN(2,XPL,YPL)
C draw short line at tau=tbar
      XPL(1)=TBAR
      XPL(2)=TBAR
      YPL(1)=.52D0
      YPL(2)=.48D0
      CALL GRSTCL(4)
      CALL GRPLIN(2,XPL,YPL)
C draw likelihood function -(l-lmax)
      FMIN=DBLE(N)*(1.+LOG(TBAR))
      DO 20 I=1,500
        XPL(I)=DBLE(I)*0.01D0
        YPL(I)=DBLE(N)*(TBAR/XPL(I)+LOG(XPL(I)))-FMIN
20    CONTINUE
      CALL GRPLIN(500,XPL,YPL)
C compute asymmetric errors
      X0=.001D0
      X1=TBAR
      CALL AUXZBR(X0,X1,FUE1ML,TBAR,N,NDUM)
      CALL AUXZFN(X0,X1,SIGMNS,FUE1ML,TBAR,N,NDUM,1.D-5)
      SIGMNS=TBAR-SIGMNS
      X0=TBAR
      X1=5.D0
      CALL AUXZBR(X0,X1,FUE1ML,TBAR,N,NDUM)
      CALL AUXZFN(X0,X1,SIGPLS,FUE1ML,TBAR,N,NDUM,1.D-5)
      SIGPLS=SIGPLS-TBAR
C compute symmetric error
      SIG=TBAR/SQRT(DBLE(N))
C prepare and draw caption
      CAPT(1:39)='N =****, t^^"-# = ****, &D@_-# = *****,'
      CAPT(40:71)=' &D@_+# = *****, &Dt^"~# = *****'
      WRITE(CAPT(4:7),'(I4)')N
      WRITE(CAPT(19:23),'(F4.2)')TBAR
      WRITE(CAPT(34:38),'(F5.2)')SIGMNS
      WRITE(CAPT(50:54),'(F5.2)')SIGPLS
      WRITE(CAPT(67:71),'(F5.2)')SIG
      CALL GRSTCL(1)
      CALL GRTXTC(1.D0,CAPT,71)
      CALL GRCLSE
      END
C-----------------------------------------------------------
      DOUBLE PRECISION FUNCTION FUE1ML(TAU,TBAR,N,NDUM)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      FUE1ML=DBLE(N)*(TBAR/TAU+LOG(TAU/TBAR)-1.D0)-.5D0
      END

File: runme/mcgen/local/datan/exasrc/e1ml.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern double CBFUNC fue1ml(double const P_T *tau, double const P_T *tbar,
                            integer const P_T *n, integer const P_T *ndum);

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__4 = 4, c__11 = 11, c__71 = 71,
               c__500 = 500;
static double c_d0 = 0., c_d5 = 5., c_dp6 = .6, c_dp1 = .1, c_d1p1 = 1.1, 
              c_dp75 = .75, c_dmp214 = -.214, c_d1p2 = 1.2, c_dmp1 = -.1, 
              c_dp9 = .9, c_d1em5 = 1e-5, c_d1 = 1.;

main(void)
{
    /* Initialized data */ /* Local variables */
    static integer ndum, i, n, iseed1, iseed2, nws;
    static double tbar, fmin, t[1000], x0, x1, sigmns, sigpls, sig,
                  xpl[500], ypl[500];
    static char capt[75], tx[75] =
                "t,&t                                              "
	        "                         ",
                ty[75] = "-(l-l_max#)                                       "
	        "                         ";
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program E1ML computes mean life and");
    printf("%s\n\n", "asymmetric errors from few radioactive decays.");
/* ask for input */
    printf("%s\n", "Enter number N of decays (1 <= N <= 1000):");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter seed number ISEED1 (>0):");
    printf("%s", "> ");
    scanf("%li", &iseed1);
    printf("%s\n", "Enter seed number ISEED2 (>0):");
    printf("%s", "> ");
    scanf("%li", &iseed2);
/* initialization */
    if (iseed1 <= 0) {
	iseed1 = 123456;
    }
    if (iseed2 <= 0) {
	iseed2 = 654321;
    }
    rne2in_(&iseed1, &iseed2);
/* prepare graphics */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
    gropen_();
    gropws_(&nws);
    grwncc_(&c_d0, &c_d5, &c_d0, &c_dp6);
    grvwwc_(&c_dp1, &c_d1p1, &c_dp1, &c_dp75);
    grwnwc_(&c_dmp214, &c_d1p2, &c_dmp1, &c_dp9);
    grstfr_(&nws, &c_d5, &c_d0);
    grfram_();
    grsclx_(tx, &c__4);
    grscly_(ty, &c__11);
    grboun_();
/* generate sample and draw 1D scatter diagram */
    rnecuy_(t, &n);
    tbar = 0.;
    ypl[0] = 0.;
    ypl[1] = .1;
    grstcl_(&c__2);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
	t[i - 1] = -log(t[i - 1]);
	tbar += t[i - 1];
	xpl[0] = t[i - 1];
	xpl[1] = t[i - 1];
	grplin_(&c__2, xpl, ypl);
    }
/* draw line -(l-lmax)=.5 */
    tbar /= (double) n;
    xpl[0] = 0.;
    xpl[1] = 5.;
    ypl[0] = .5;
    ypl[1] = .5;
    grplin_(&c__2, xpl, ypl);
/* draw short line at tau=tbar */
    xpl[0] = tbar;
    xpl[1] = tbar;
    ypl[0] = .52;
    ypl[1] = .48;
    grstcl_(&c__4);
    grplin_(&c__2, xpl, ypl);
/* draw likelihood function -(l-lmax) */
    fmin = (double) n * (log(tbar) + 1.f);
    for (i = 1; i <= 500; ++i) {
	xpl[i - 1] = (double) i * .01;
	ypl[i - 1] = (double) n * (tbar / xpl[i - 1] + log(xpl[i - 1])) - fmin;
    }
    grplin_(&c__500, xpl, ypl);
/* compute asymmetric errors */
    x0 = .001;
    x1 = tbar;
    auxzbr_(&x0, &x1, fue1ml, &tbar, &n, &ndum);
    auxzfn_(&x0, &x1, &sigmns, fue1ml, &tbar, &n, &ndum, &c_d1em5);
    sigmns = tbar - sigmns;
    x0 = tbar;
    x1 = 5.;
    auxzbr_(&x0, &x1, fue1ml, &tbar, &n, &ndum);
    auxzfn_(&x0, &x1, &sigpls, fue1ml, &tbar, &n, &ndum, &c_d1em5);
    sigpls -= tbar;
/* compute symmetric error */
    sig = tbar / sqrt((double) n);
/* prepare and draw caption */
    sprintf(capt, "N =%4li, t^^\"-# = %4.2lf, &D@_-# = %5.2lf,"
            " &D@_+# = %5.2lf, &Dt^\"~# = %5.2lf",
            n, tbar, sigmns, sigpls, sig);
    memset(capt + 71, ' ', 4);
    grstcl_(&c__1);
    grtxtc_(&c_d1, capt, &c__71);
    grclse_();
    return 0;
} /* main */

/* ----------------------------------------------------------- */
double CBFUNC fue1ml(double const P_T *tau, double const P_T *tbar,
                     integer const P_T *n, integer const P_T *ndum)
{
    /* System generated locals */
    double ret_val;

    ret_val = (double) (*n) * (*tbar / *tau + log(*tau / *tbar) - 1.) - .5;
    return ret_val;
} /* fue1ml */

